### PR TITLE
Add layer_map_analyzer tool

### PR DESCRIPTION
**Pull request by knizhnik**
See #3348

### CODE REVIEW

**Comment by koivunej** [comment]
@knizhnik did you forget to add the new binary, diffs are all about making items accessible?

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    let holes = Vec::new();
+    Some(LayerFile {
+        key_range,
+        lsn_range,
+        is_delta,
+        holes,
+    })
+}
+
+fn get_holes(path: &Path, max_holes: usize) -> Result<Vec<Hole>> {
```
Comment (bojanserafimov) [inline pageserver/src/bin/layer_map_analyzer.rs:91]: nit: let's document the behavior: "Finds the `max_holes` largest holes, ignoring any that are smaller than `MIN_HOLE_LENGTH`"

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 2]
```diff
@@ -0,0 +1,212 @@
+use anyhow::Result;
```
Comment (bojanserafimov) [inline pageserver/src/bin/layer_map_analyzer.rs:1]: Top-level docstring would be useful. Suggestion:

"
Tool for extracting content-dependent metadata about layers. Useful for scanning real project layer files and evaluating the effectiveness of different heuristics on them.

Currently it only analyzes holes, which are regions within the layer range that the layer contains no updates for. In the future it might do more analysis (maybe key quantiles?) but it should never return sensitive data.
"

Also maybe `layer_metadata_analyzer` is another alternative name. The current name makes it sound like the input should be a layer map. Your choice.

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        return Ok(());
+    }
+    let storage_path = PathBuf::from_str(&args[1])?;
+    let max_holes = if args.len() > 2 {
+        args[2].parse::<usize>().unwrap()
+    } else {
+        DEFAULT_MAX_HOLES
+    };
+    pageserver::virtual_file::init(10);
+    pageserver::page_cache::init(100);
```
Comment (bojanserafimov) [inline pageserver/src/bin/layer_map_analyzer.rs:138]: why are these needed? Maybe a comment would help

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        }
+        for timeline in fs::read_dir(tenant.path().join("timelines"))? {
+            let timeline = timeline?;
+            if !timeline.file_type()?.is_dir() {
+                continue;
+            }
+            let mut layers = Vec::new();
+            let mut n_deltas = 0usize;
+            let mut n_excess_layers = 0usize;
+            let mut n_holes = 0usize;
```
Comment (bojanserafimov) [inline pageserver/src/bin/layer_map_analyzer.rs:155]: Let's declare these before the next for loop because that's when we start computing them. Mutable variables inside a long scope are confusing.

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        let tenant = tenant?;
+        if !tenant.file_type()?.is_dir() {
+            continue;
+        }
+        for timeline in fs::read_dir(tenant.path().join("timelines"))? {
+            let timeline = timeline?;
+            if !timeline.file_type()?.is_dir() {
+                continue;
+            }
+            let mut layers = Vec::new();
```
Comment (bojanserafimov) [inline pageserver/src/bin/layer_map_analyzer.rs:152]:
```suggestion
            // Collect sorted vec of layers and count deltas
            let mut layers = Vec::new();
```

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                    if layer_file.is_delta {
+                        layer_file.holes = get_holes(&layer.path(), max_holes)?;
+                        n_deltas += 1;
+                    }
+                    layers.push(layer_file);
+                }
+            }
+            layers.sort_by_key(|layer| layer.lsn_range.end);
+
+            for i in 0..layers.len() {
```
Comment (bojanserafimov) [inline pageserver/src/bin/layer_map_analyzer.rs:170]:
```suggestion
            // Count the number of holes and number of excess layers
            for i in 0..layers.len() {
```

Maybe also define what an excess layer is

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                    let img_key_range = &layers[i].key_range;
+                    for j in (0..i).rev() {
+                        if range_overlaps(img_key_range, &layers[j].key_range) {
+                            if layers[j].is_delta {
+                                n_deltas_since_last_image += 1;
+                                if layers[j].skips(img_key_range) {
+                                    n_skipped += 1;
+                                }
+                            } else {
+                                break;
```
Comment (bojanserafimov) [inline pageserver/src/bin/layer_map_analyzer.rs:183]: This isn't technically correct, since this image might not cover the full key range. Maybe it's a good approximation but let's put a `// HACK` comment to make it clear.

There's another approximation: We should also check if this image is excess or not. If it's excess we can't break because the image layer wouldn't have existed if we used the new heuristic. **Your call whether it's worth fixing or just documenting.**

And there's yet another: The new layer map doesn't count total deltas, but the max stack of overlapping deltas.

**Review by bojanserafimov — APPROVED** [review]
Good stuff!

At some point I'll try integrating this into the visualizer too, so we can see the occupied ranges.

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 7 above)
Comment (bojanserafimov) [reply pageserver/src/bin/layer_map_analyzer.rs:183]: But probably long term we'd want the real layer_map+reimaging+compaction code to work only with metadata so that we can unit test it and use it here without rewriting it. It's not a small task though.

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+use pageserver::tenant::storage_layer::delta_layer::{Summary, DELTA_KEY_SIZE};
+use pageserver::tenant::storage_layer::range_overlaps;
+use pageserver::virtual_file::VirtualFile;
+
+use utils::{bin_ser::BeSer, lsn::Lsn};
+
+const MIN_HOLE_LENGTH: i128 = (128 * 1024 * 1024 / PAGE_SZ) as i128;
+const DEFAULT_MAX_HOLES: usize = 10;
+
+/// Wrapper for key range to provide reverse ordering by range length for BinaryHeap
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:20]: Sorry, but `Reverse` can not e used here because I am sorting ranges by length and it requires custom comparator in any case.

File: pageserver/src/bin/layer_map_analyzer.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+    let is_delta = if lsns.len() == 1 {
+        lsns.push(lsns[0]);
+        false
+    } else {
+        true
+    };
+
+    let key_range = Key::from_hex(keys[0]).unwrap()..Key::from_hex(keys[1]).unwrap();
+    let lsn_range = Lsn::from_hex(lsns[0]).unwrap()..Lsn::from_hex(lsns[1]).unwrap();
+    let holes = Vec::new();
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:82]: Sorry, but this code was cut&paste of your draw_timeline_dir.rs
Concerning initialization of `holes` - it will be filled later for delta layers.

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:91]: This function was added to Layers trait in #3348

Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:91]: Done

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 2 above)
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:1]: 1. Thank you for writing this description for me. Added.
2. Right now holes is not part of layer metadata. It will be if #3348 will be merged.
3.

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 3 above)
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:138]: Them are needed to access layer persistent B-Tree.
Comment is added.

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 4 above)
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:155]: Done

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 5 above)
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:152]: Done

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 6 above)
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:170]: Done

File: pageserver/src/bin/layer_map_analyzer.rs (hunk 7 above)
Comment (knizhnik) [reply pageserver/src/bin/layer_map_analyzer.rs:183]: Well, right now image layer is always dense, despite to the fact that it doesn't contain all possible key values in the specified range: there are may be no keys in the storage belonging to the image layer range but not present in the image layer. 

Comments added.
